R/constraints.R

```diff
@@ -0,0 +1,133 @@
+#' Formulating constraints
+#'
+#' [TODO]
+#'
+#' [TODO: currently we only support scores on the LHS - this can be easily extended!]
+#'
+#' [TODO: do we also want to support scores vs scores comparisons?]
+#'
+#' @param e1 first comparator
+#' @param e2 second comparator
+#' @template dotdotdotTemplate
+#'
+#' @exportClass Constraint
+setClass("Constraint")
+
+#' @param s constraint to evaluate [TODO make naming of arguments for evaluate more generic!]
+#' @param design TwoStageDesign to evaluate
+#'
+#' @rdname Constraint-class
+#' @export
+setMethod("evaluate", signature("Constraint", "TwoStageDesign"),
+          function(s, design, ...) evaluate(s@score, design, ...) - s@rhs )
+
+
+
+
+
+#' @rdname Constraint-class
+#' @exportClass ConditionalConstraint
+setClass("ConditionalConstraint", representation(
+        score = "AbstractConditionalScore",
+        rhs   = "numeric"
+    ),
+    contains = "Constraint")
+
+
+#' @rdname Constraint-class
+#' @export
+setMethod("<=", signature("AbstractConditionalScore", "numeric"),
```

Comment:
this is where the magic happens - by overloading the <= operator we can create constraints very naturally!